src/cartservice/cartstore/RedisCartStore.cs

```diff
@@ -115,86 +117,95 @@ public async Task AddItemAsync(string userId, string productId, int quantity)
                     }
                     else
                     {
-                        existingItem.Quantity += quantity;
+                        cart = Hipstershop.Cart.Parser.ParseFrom(value);
+                        var existingItem = cart.Items.SingleOrDefault(i => i.ProductId == productId);
+                        if (existingItem == null)
+                        {
+                            cart.Items.Add(new Hipstershop.CartItem { ProductId = productId, Quantity = quantity });
+                        }
+                        else
+                        {
+                            existingItem.Quantity += quantity;
+                        }
                     }
-                }
 
-                await db.HashSetAsync(userId, new[]{ new HashEntry(CART_FIELD_NAME, cart.ToByteArray()) });
-            }
-            catch (Exception ex)
-            {
-                throw new RpcException(new Status(StatusCode.FailedPrecondition, $"Can't access cart storage. {ex}"));
+                    await db.HashSetAsync(userId, new[] { new HashEntry(CART_FIELD_NAME, cart.ToByteArray()) });
+                }
+                catch (Exception ex)
+                {
+                    throw new RpcException(new Status(StatusCode.FailedPrecondition, $"Can't access cart storage. {ex}"));
+                }
             }
         }
 
         public async Task EmptyCartAsync(string userId)
         {
             Console.WriteLine($"EmptyCartAsync called with userId={userId}");
 
-            try
+            using (var parent = CartActivity.ActivitySource.StartActivity("EmptyCart", ActivityKind.Server))
             {
-                EnsureRedisConnected();
-                var db = redis.GetDatabase();
+                try
+                {
+                    EnsureRedisConnected();
+                    var db = redis.GetDatabase();
 
-                // Update the cache with empty cart for given user
-                await db.HashSetAsync(userId, new[] { new HashEntry(CART_FIELD_NAME, emptyCartBytes) });
-            }
-            catch (Exception ex)
-            {
-                throw new RpcException(new Status(StatusCode.FailedPrecondition, $"Can't access cart storage. {ex}"));
+                    // Update the cache with empty cart for given user
+                    await db.HashSetAsync(userId, new[] { new HashEntry(CART_FIELD_NAME, emptyCartBytes) });
+                }
+                catch (Exception ex)
+                {
+                    throw new RpcException(new Status(StatusCode.FailedPrecondition, $"Can't access cart storage. {ex}"));
+                }
             }
         }
 
         public async Task<Hipstershop.Cart> GetCartAsync(string userId)
         {
             Console.WriteLine($"GetCartAsync called with userId={userId}");
 
-            try
+            using (var parent = CartActivity.ActivitySource.StartActivity("GetCart", ActivityKind.Server))
```

Comment:
Small nit, but by using the following syntax, you can reduce the changes you've introduced with the additional nesting. May have the benefit of making future merges from the upstream repo less of a headache.

```csharp
using var parent = CartActivity.ActivitySource.StartActivity("GetCart", ActivityKind.Server);
```

Reply:
Might need to up the `LangVersion` for the project...

Reply:
Okay updated, that syntax is new to me. Does this just go out of scope at the end of the method?

Reply:
Yep, `parent.Dispose()` will occur at the end of the scope in which `parent` was declared. I kind of have a love/hate relationship with this syntax. Nice that it reduces nesting, however, it is far to easy to leave off the `using` - under many circumstances things would continue to compile and run fine though you wouldn't be disposing things that maybe you had intended to.